### PR TITLE
feat: support `map_access`

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?
part of #17743

- test `to_jsonb(map)`. This was already supported at the beginning (via `ToJsonb` trait)
- support `map_access`. The main work is about type inference.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by xxchan** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17986?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#17986** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17986?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#17980** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17980?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#17860** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17860?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#17864](https://github.com/risingwavelabs/risingwave/pull/17864) <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17864?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#17690** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17690?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#17763** <a href="https://app.graphite.dev/github/pr/risingwavelabs/risingwave/17763?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @xxchan and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: src/expr/impl/src/scalar/array.rs [hunk 1]
```diff
@@ -14,16 +14,20 @@
...
+use risingwave_common::types::{
+    DataType, ListRef, MapRef, MapType, MapValue, ScalarRefImpl, ToOwnedDatum,
+};
 use risingwave_expr::expr::Context;
 use risingwave_expr::{function, ExprError};
 
-#[function("array(...) -> anyarray", type_infer = "panic")]
+use super::array_positions::array_position;
+
+#[function("array(...) -> anyarray", type_infer = "unreachable")]
```
Comment (chenzl25) [inline src/expr/impl/src/scalar/array.rs:25]: What's the different between type_infer = "panic" and "unreachable"?

**Review by chenzl25 — APPROVED** [review]
LGTM.  BTW, DuckDB seems to have a more user friendly syntax for map type. I think we can create a syntax sugar for them as well.
```
SELECT MAP {'key1': 5, 'key2': 43}['key1'];
```

File: src/expr/impl/src/scalar/array.rs (hunk 1 above)
Comment (xxchan) [reply src/expr/impl/src/scalar/array.rs:25]: I renamed `panic` to `unreachable` in this PR (`src/expr/macro/src/gen.rs`). You can check the panic message there.

**Comment by xxchan** [comment]
> DuckDB seems to have a more user friendly syntax for map type. I think we can create a syntax sugar for them as well.

Yes, the literal and operator syntax is on the plan, and tracked in https://github.com/risingwavelabs/risingwave/issues/17743

**Pull request merged**